src/filetest/binemit.rs

```diff
@@ -104,16 +104,54 @@ impl SubTest for TestBinEmit {
         // value locations. The current error reporting is just crashing...
         let mut func = func.into_owned();
 
+        let is_compressed = isa.flags().is_compressed();
+
         // Give an encoding to any instruction that doesn't already have one.
         for ebb in func.layout.ebbs() {
             for inst in func.layout.ebb_insts(ebb) {
                 if !func.encodings[inst].is_legal() {
-                    if let Ok(enc) = isa.encode(
+                    let mut legal_encodings = isa.legal_encodings(
                         &func.dfg,
                         &func.dfg[inst],
                         func.dfg.ctrl_typevar(inst),
-                    )
-                    {
+                    );
+
+                    let enc = if is_compressed {
+                        // Get the smallest legal encoding
+                        legal_encodings
+                            .filter(|e| {
+                                let mut result = true;
+                                let constraints = &encinfo.constraints[e.recipe()];
+
+                                for (&arg, constraint) in
+                                    func.dfg.inst_args(inst).iter().zip(constraints.ins)
+                                {
+                                    let reg = func.locations[arg].unwrap_reg();
+                                    if !constraint.regclass.contains(reg) {
```

Comment:
You have the right idea here, but this code would be better if it were broken down into smaller functions.

There are more kinds of constraints that just register classes, and you can't assume that a value location is a register. Values can also live in stack slots, and there will be instruction encodings that can access stack slots directly.

To begin with, you can just match a `ValueLoc::Stack` against a `ConstraintKind::Stack`. I will be adding some complexity to this type of constraint soon, though.